### PR TITLE
feat: add support for requiring basic finalizers

### PR DESCRIPTION
**Pull request by KevinEady**
Introduce `NODE_ADDON_API_REQUIRE_BASIC_FINALIZERS` preprocessor directive, which adds an always-fail static assertion inside the `Wrapper` definition (responsible for passing the user's non-basic finalizer through a `node_api_post_finalizer` call) if this preprocessor directive is defined.

### CODE REVIEW

File: doc/finalization.md [hunk 1]
```diff
@@ -8,7 +8,9 @@ provide more efficient memory management, optimizations, improved execution, or
 other benefits.
 
 In general, it is best to use basic finalizers whenever possible (eg. when
-access to JavaScript is _not_ needed).
+access to JavaScript is _not_ needed). To ensure that all finalizers are basic
+finalizers at compile-time, define the `NODE_ADDON_API_REQUIRE_BASIC_FINALIZERS`
```
Comment (legendecas) [inline doc/finalization.md:12]: I am wondering if we should add a "best practice" target that enables all possible checks in https://github.com/nodejs/node-addon-api/blob/main/node_addon_api.gyp.

File: doc/finalization.md [hunk 2]
```diff
@@ -8,7 +8,9 @@ provide more efficient memory management, optimizations, improved execution, or
 other benefits.
 
 In general, it is best to use basic finalizers whenever possible (eg. when
-access to JavaScript is _not_ needed).
+access to JavaScript is _not_ needed). To ensure that all finalizers are basic
```
Comment (gabrielschulhof) [inline doc/finalization.md:11]: Let's make this passive voice: The `NODE_ADDON_API_REQUIRE_BASIC_FINALIZERS` preprocessor directive can be defined to ensure that all finalizers are basic.

File: doc/setup.md [hunk 3]
```diff
@@ -90,3 +90,13 @@ provide feedback to the user of the runtime error, as it is impossible to pass
 the error to JavaScript when the environment is terminating. In order to bypass
 this behavior such that the Node process will not terminate, define the
```
Comment (gabrielschulhof) [inline doc/setup.md:91]: Here too: "The ... can be defined in order to bypass ... "

File: napi-inl.h [hunk 4]
```diff
@@ -205,14 +205,19 @@ struct FinalizeData {
     });
   }
 
-#ifdef NODE_API_EXPERIMENTAL_HAS_POST_FINALIZER
+#if defined(NODE_API_EXPERIMENTAL_HAS_POST_FINALIZER)
```
Comment (gabrielschulhof) [inline napi-inl.h:208]: Is this change needed?

File: napi-inl.h [hunk 5]
```diff
@@ -235,14 +240,19 @@ struct FinalizeData {
...
```
Comment (gabrielschulhof) [inline napi-inl.h:243]: Same here.

File: test/require_basic_finalizers/index.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    cwd: ADDON_FOLDER
+  }), 'Addon unexpectedly compiled successfully');
+
+  // Succeed when NODE_ADDON_API_REQUIRE_BASIC_FINALIZERS is not enabled
+  return assert.doesNotReject(exec('npm install', {
+    cwd: ADDON_FOLDER
+  }));
+}
+
+module.exports = (function () {
```
Comment (gabrielschulhof) [inline test/require_basic_finalizers/index.js:29]: Do we need to assign to module.exports here? Could we simply execute the contents of this IIFE at the top level?

**Review by gabrielschulhof — APPROVED** [review]
LGTM with some comments.

File: napi-inl.h (hunk 4 above)
Comment (KevinEady) [reply napi-inl.h]: Whoops, this was a leftover of a different implementation. Addressed in https://github.com/nodejs/node-addon-api/pull/1568/commits/23b57cc93946df51f30cbfada61213be14f61421

File: napi-inl.h (hunk 5 above)
Comment (KevinEady) [reply napi-inl.h]: Addressed in https://github.com/nodejs/node-addon-api/pull/1568/commits/23b57cc93946df51f30cbfada61213be14f61421

File: doc/finalization.md (hunk 2 above)
Comment (KevinEady) [reply doc/finalization.md]: Addressed in https://github.com/nodejs/node-addon-api/pull/1568/commits/23b57cc93946df51f30cbfada61213be14f61421

File: doc/setup.md (hunk 3 above)
Comment (KevinEady) [reply doc/setup.md]: Addressed in https://github.com/nodejs/node-addon-api/pull/1568/commits/23b57cc93946df51f30cbfada61213be14f61421

**Comment by KevinEady** [comment]
> LGTM with some comments.

Hi @gabrielschulhof , 

Your changes have been introduced in https://github.com/nodejs/node-addon-api/pull/1568/commits/23b57cc93946df51f30cbfada61213be14f61421. PTAL!

**Review by legendecas — APPROVED** [review]

**Review by mhdawson — APPROVED** [review]
LGTM

**Review by gabrielschulhof — APPROVED** [review]
